input/postgres/explain.go

```diff
@@ -44,40 +44,53 @@ func RunExplain(server state.Server, inputs []state.PostgresQuerySample, collect
 			logger.PrintVerbose("Could not connect to %s to run explain: %s; skipping", dbName, err)
 			continue
 		}
+		useHelper := statsHelperExists(db, "explain")
+		if useHelper {
+			logger.PrintVerbose("Found pganalyze.explain() stats helper in database \"%s\"", dbName)
+		} else if !connectedAsSuperUser(db, server.Config.SystemType) {
+			logger.PrintInfo("Warning: pganalyze.explain() helper function not found in database \"%s\". Please set up"+
+				" the monitoring helper functions (https://github.com/pganalyze/collector#setting-up-a-restricted-monitoring-user)"+
+				" in every database you want to monitor to avoid permissions issues when running log-based EXPLAIN.", dbName)
+		}
 
-		dbOutputs := runDbExplain(db, dbSamples)
+		dbOutputs := runDbExplain(db, dbSamples, useHelper)
 		db.Close()
 
 		outputs = append(outputs, dbOutputs...)
 	}
 	return
 }
 
-func runDbExplain(db *sql.DB, inputs []state.PostgresQuerySample) (outputs []state.PostgresQuerySample) {
+func runDbExplain(db *sql.DB, inputs []state.PostgresQuerySample, useHelper bool) (outputs []state.PostgresQuerySample) {
 	for _, sample := range inputs {
 		// To be on the safe side never EXPLAIN a statement that can't be parsed,
 		// or multiple statements in one (leading to accidental execution)
 		parsetree, err := pg_query.Parse(sample.Query)
-		if err == nil && len(parsetree.Statements) == 1 {
-			stmt := parsetree.Statements[0].(pg_query_nodes.RawStmt).Stmt
-			switch stmt.(type) {
-			case pg_query_nodes.SelectStmt, pg_query_nodes.InsertStmt, pg_query_nodes.UpdateStmt, pg_query_nodes.DeleteStmt:
-				sample.HasExplain = true
-				sample.ExplainSource = pganalyze_collector.QuerySample_STATEMENT_LOG_EXPLAIN_SOURCE
-				sample.ExplainFormat = pganalyze_collector.QuerySample_JSON_EXPLAIN_FORMAT
-
+		if err != nil || len(parsetree.Statements) != 1 {
+			continue
+		}
+		stmt := parsetree.Statements[0].(pg_query_nodes.RawStmt).Stmt
+		switch stmt.(type) {
+		case pg_query_nodes.SelectStmt, pg_query_nodes.InsertStmt, pg_query_nodes.UpdateStmt, pg_query_nodes.DeleteStmt:
+			sample.HasExplain = true
+			sample.ExplainSource = pganalyze_collector.QuerySample_STATEMENT_LOG_EXPLAIN_SOURCE
+			sample.ExplainFormat = pganalyze_collector.QuerySample_JSON_EXPLAIN_FORMAT
+
+			if useHelper {
+				err = db.QueryRow(QueryMarkerSQL+"SELECT pganalyze.explain($1, $2)", sample.Query, sample.Parameters).Scan(&sample.ExplainOutput)
```

Comment:
Looks like this needs `pq.Array`, otherwise we get an error (`sql: converting argument $2 type: unsupported type []string, a slice of string`)

```diff
diff --git a/input/postgres/explain.go b/input/postgres/explain.go
index 22d7ca7..91a20fb 100644
--- a/input/postgres/explain.go
+++ b/input/postgres/explain.go
@@ -77,7 +77,7 @@ func runDbExplain(db *sql.DB, inputs []state.PostgresQuerySample, useHelper bool
                        sample.ExplainFormat = pganalyze_collector.QuerySample_JSON_EXPLAIN_FORMAT
 
                        if useHelper {
-                               err = db.QueryRow(QueryMarkerSQL+"SELECT pganalyze.explain($1, $2)", sample.Query, sample.Parameters).Scan(&sample.ExplainOutput)
+                               err = db.QueryRow(QueryMarkerSQL+"SELECT pganalyze.explain($1, $2)", sample.Query, pq.Array(sample.Parameters)).Scan(&sample.ExplainOutput)
                                if err != nil {
                                        sample.ExplainError = fmt.Sprintf("%s", err)
                                }
```

Reply:
Nice catch, I hadn't actually tried that change yet.